week3/homework/src/lib.js

```diff
@@ -0,0 +1,131 @@
+/* eslint-disable max-len */
+const fs = require('fs');
+const { promisify } = require('util');
+const uuidv4 = require('uuid/v4');
+const Joi = require('@hapi/joi');
+
+class Todos {
+  constructor(file) {
+    this.file = file;
+    this.readFromFile = promisify(fs.readFile).bind(this, this.file, 'utf8');
+    this.writeToFile = promisify(fs.writeFile).bind(this, this.file);
+  }
+
+  async readTodos(req, res) {
+    try {
+      const todos = JSON.parse(await this.readFromFile());
+      res.json(todos);
+    } catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async readTodo(req, res) {
+    try {
+      const todos = JSON.parse(await this.readFromFile());
+      const todo = todos.find(todo => todo.id === req.params.id);
+      if (!todo) return res.status(404).send(`${req.params.id} is not a valid id`);
+      res.json(todo);
+    } catch (error) {
+      Todos.handleError(res, error);
+    }
+  }
+
+  async createTodos(req, res) {
+    try {
+      const { error } = Todos.validateTodo(req.body);
+      if (error) return res.status(400).send(error.details.map(d => d.message).join('\n'));
+      const { todo } = req.body;
+      const todos = JSON.parse(await this.readFromFile());
+      todos.push(Object.assign(todo, { id: uuidv4(), done: false }));
```

Comment:
Personally always using spread operator with arrays. Thanks for the tip 

> `Object.assign()` modifies an object in place, and so it can trigger ES6 setters. If you prefer using immutable techniques, the object spread operator is a clear winner.